packages/salesforcedx-vscode-soql/src/queryDataView/queryDataFileService.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import { JsonMap } from '@salesforce/ts-types';
+import * as fs from 'fs';
+import { QueryResult } from 'jsforce';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { getRootWorkspacePath } from '../commonUtils';
+import { QUERY_RESULTS_DIR_NAME, QUERY_RESULTS_DIR_PATH } from '../constants';
+import {
+  CsvDataProvider,
+  DataProvider,
+  JsonDataProvider
+} from './dataProviders';
+
+export enum FileFormat {
+  JSON = 'json',
+  CSV = 'csv'
+}
+
+export class QueryDataFileService {
+  private dataProvider: DataProvider;
+
+  constructor(
+    private queryData: QueryResult<JsonMap>,
+    private format: FileFormat,
+    private documentName: string
+  ) {
+    this.dataProvider = this.getDataProvider();
+  }
+
+  protected getDataProvider(): DataProvider {
+    switch (this.format) {
+      case FileFormat.CSV:
+        return new CsvDataProvider(this.documentName);
+      case FileFormat.JSON:
+        return new JsonDataProvider(this.documentName);
+      default:
+        throw new Error('No DataProvider Found');
+    }
+  }
+
+  public save(): string {
+    try {
+      const fileContent = this.dataProvider.getFileContent(
+        this.queryData.records
+      );
+      const savedFileName = this.dataProvider.getFileName();
+      const queryDataFilePath = path.join(
+        this.getResultsDirectoryPath(),
+        savedFileName
+      );
+
+      this.createResultsDirectoryIfDoesNotExist();
+      // Save query results to disk
+      fs.writeFileSync(queryDataFilePath, fileContent);
```

Comment:
Consider possibly using `vscode.workspace.fs` methods for fs operations.

Reply:
I was trying to decide wether to use the VS code fs API actually, but I did not see a distinct advantage, in fact it seemed more verbose because I had to convert everything to vs code URI etc.

Is there an advantage to using `vscode.workspace.fs`?

Ultimately though, I decided to use the regular file system module because our goal is to build this as both a stand alone web app and VS code extension, so we want to limit any dependency on VS code APIs

Reply:
The main advantage is that it's flexible with the file system being used, whether it's local or some remote container depending on the context the extensions are loaded in. If it's going to be a standalone app at some point the NodeJS fs dependency will have to be removed as well. I don't think it matters either way at this point, but worth considering.

Reply:
Oh yeah, good point about the node dependency, node does not equal browser, ha. We aren't sure what the standalone app will look like yet, we'll cross that bridge when we get there.